accepted/0000-missing-command-lookups.md

```diff
@@ -0,0 +1,70 @@
+# Support Yarn style command/script/bin lookups from the CLI
+
+## Summary
+
+Allow for a CLI experience like:
+
+```sh
+npm install typescript
+npm tsc
+```
+
+## Motivation
+
+I have two main motivations:
+
+- Personal, looking at the list of yarn commands I [use in an average month](https://gist.github.com/orta/489dbfa1b7467804955ea7800adf1494#file-last_month_yarn-txt) (1200) roughly 30% (400) of those commands are using the shorthand syntax to run a command installed into `node_modules/.bin` in a project.
+
+- Professional, it'd be great for the TypeScript docs to be able to simply say: `npm install typescript --save-dev' then `npm tsc index.ts` and there to be no ambiguity like our current recommendation for `npx tsc index.ts`.
+
+## Detailed Explanation
+
+When npm receives a command it does not recognise, npm first resolves the next string to see if it matches:
+
+ - Available "scripts" in `package.json`
+ - Commands matching in `./node_modules/.bin/`
+
+Failing that, then show help.
+
+## Rationale and Alternatives
+
+1. Having a `script`s section like this:
+
+    ```json
+    {
+      "scripts": {
+        "tsc": "tsc",
+        "eslint": "eslint",
+        "prettier": "prettier"
+      }
+    }
+    ```
+
+    Is a waste of space in the JSON, the `node_modules/.bin` folder is typically full of commands and after a while you either start prefixing commands with `node_modules/.bin` instead or you fill up the scripts section.
+
+2. This change removes the concept of "blessed" scripts. `npm test` works but `npm test:lint` does not. Eventually folks memorize the ones they can use shorthand for, but there's no reason for the distinction other than at some point it was decided to be given a short-cut. These special cases can be removed with this change.
+
+3. It does not break any existing setups. 
```

Comment:
if npm added an "npm build" command the release after shipping this RFC, though, it would break a great many existing setups.

Reply:
This is a possibility yes, at that point there would probably be a major npm semver and breaking changes like that can be handled.

Reply:
Right - but the consequence is that npm could no longer add top-level commands without it being a semver-major change; and regardless of semver-major _allowing_ them to do it, that doesn't change the disruption it might cause for users.

Reply:
IMO that's worth it. This behavior accounts for roughly 40% of my command line usage in yarn. I don't maintain npm, so that's just a anecdata but this feature has been a constant win for yarn from day 1. Even post npm 7 and in yarn deprecated land, I regularly recommend people to not use npm just for this alone.

There's possible mitigation strategies like warnings when they match a potential upcoming command, or just outright allowing the script to take precedence over _new_ commands. npm could emit that this is blocking a known CLI command and recommend a rename while still running the same script. I am assuming under the hood a new `npm publish2` isn't directly calling out to the shell with `npm pack2` for example, and so adding a script `pack2` wouldn't take precedence in the internals and people's projects wouldn't break.